File: web/documentserver-example/python/src/proxy/proxy.py
#  Copyright Ascensio System SIA 2024
#  SPDX-FileCopyrightText: 2024 Ascensio System SIA
#
#  SPDX-License-Identifier: Ascensio-System
#
#     Our License onlyoffice.com
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     

from urllib.parse import ParseResult
from src.configuration import ConfigurationManager


class ProxyManager():
    config_manager: ConfigurationManager

    def __init__(self, config_manager: ConfigurationManager):
        self.config_manager = config_manager

    def resolve_url(self, url: ParseResult) -> ParseResult:
        if not self.__refer_public_url(url):
            return url
        return self.__redirect_public_url(url)

    def __refer_public_url(self, url: ParseResult) -> bool:
        public_url = self.config_manager.document_server_public_url()
        return (
            url.scheme == public_url.scheme and
            url.hostname == public_url.hostname and
            url.port == public_url.port
        )

    def __redirect_public_url(self, url: ParseResult) -> ParseResult:
        private_url = self.config_manager.document_server_private_url()
        return ParseResult(
            scheme=private_url.scheme,
            netloc=private_url.netloc,
            path=url.path,
            params=url.params,
            query=url.query,
            fragment=url.fragment
        )

File: web/documentserver-example/ruby/app/proxy/proxy.rb
#  Copyright Ascensio System SIA 2024
#  SPDX-FileCopyrightText: 2024 Ascensio System SIA
#
#  SPDX-License-Identifier: Ascensio-System
#
#     Our License onlyoffice.com
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     Empty line
#     

# frozen_string_literal: true
# typed: true

require 'sorbet-runtime'
require 'uri'
require_relative '../configuration/configuration'

# Class manages URI resolution, redirecting public URLs to private ones based on the configuration.
class ProxyManager
  extend T::Sig

  sig { params(config_manager: ConfigurationManager).void }
  def initialize(config_manager:)
    @config_manager = config_manager
  end

  sig { params(uri: URI::Generic).returns(URI::Generic) }
  def resolve_uri(uri)
    return uri unless refer_public_url(uri)

    redirect_public_url(uri)
  end

  sig { params(uri: URI::Generic).returns(T::Boolean) }
  private def refer_public_url(uri)
    public_uri = @config_manager.document_server_public_uri
    uri.scheme == public_uri.scheme &&
    uri.host == public_uri.host &&
    uri.port == public_uri.port
  end

  sig { params(uri: URI::Generic).returns(URI::Generic) }
  private def redirect_public_url(uri)
    private_uri = @config_manager.document_server_private_uri
    redirected_uri = uri
    redirected_uri.scheme = private_uri.scheme
    redirected_uri.host = private_uri.host
    redirected_uri.port = private_uri.port
    redirected_uri
  end
end
